File: src/ros2_bridge/mappers/sensor_msgs/joint_state.rs
//! Typed mapper for `sensor_msgs/msg/JointState`.

use crate::ros2_bridge::mapper::TypedTopicMapper;

pub(crate) fn joint_state_to_bus(
    msg: ros_env::sensor_msgs::msg::JointState,
) -> crate::sensor_msgs::msg::v1::JointState {
    crate::sensor_msgs::msg::v1::JointState {
        header: Some(crate::ros2_bridge::mappers::std_msgs::header::header_to_bus(msg.header)),
        name: crate::ros2_bridge::mappers::convert::string_seq(msg.name),
        position: crate::ros2_bridge::mappers::convert::f64_seq(msg.position),
        velocity: crate::ros2_bridge::mappers::convert::f64_seq(msg.velocity),
        effort: crate::ros2_bridge::mappers::convert::f64_seq(msg.effort),
    }
}

pub(crate) fn joint_state_to_ros(
    bus: crate::sensor_msgs::msg::v1::JointState,
) -> ros_env::sensor_msgs::msg::JointState {
    ros_env::sensor_msgs::msg::JointState {
        header: crate::ros2_bridge::mappers::std_msgs::header::header_to_ros(
            bus.header.unwrap_or_default(),
        ),
        name: crate::ros2_bridge::mappers::convert::ros_string_seq(bus.name),
        position: crate::ros2_bridge::mappers::convert::FromF64Seq::from_f64_seq(bus.position),
        velocity: crate::ros2_bridge::mappers::convert::FromF64Seq::from_f64_seq(bus.velocity),
        effort: crate::ros2_bridge::mappers::convert::FromF64Seq::from_f64_seq(bus.effort),
    }
}

#[derive(Clone, Copy, Debug, Default)]
pub struct SensorMsgsJointStateMapper;

impl TypedTopicMapper for SensorMsgsJointStateMapper {
    type Ros = ros_env::sensor_msgs::msg::JointState;
    type Bus = crate::sensor_msgs::msg::v1::JointState;

    fn ros_to_bus(&self, msg: Self::Ros) -> crate::errors::Result<Self::Bus> {
        Ok(joint_state_to_bus(msg))
    }

    fn bus_to_ros(&self, msg: Self::Bus) -> crate::errors::Result<Self::Ros> {
        Ok(joint_state_to_ros(msg))
    }
}

File: bindings/python/robot_bus/ros2_bridge/mappers/sensor_msgs/joint_state.py
"""Generated mapper for `sensor_msgs/msg/JointState`."""

from __future__ import annotations

from robot_bus.ros2_bridge.mappers import _convert
from robot_bus.ros2_bridge.mappers.std_msgs.header import header_to_bus, header_to_ros

_BusMsg = None


def _bus_cls():
    global _BusMsg
    if _BusMsg is None:
        from robot_bus.sensor_msgs.msg.v1 import JointState as BusMsg

        _BusMsg = BusMsg
    return _BusMsg


def joint_state_to_bus(msg):
    BusMsg = _bus_cls()
    bus = BusMsg()
    bus.header.CopyFrom(header_to_bus(msg.header))
    bus.name.extend([str(x) for x in msg.name])
    bus.position.extend(list(msg.position))
    bus.velocity.extend(list(msg.velocity))
    bus.effort.extend(list(msg.effort))
    return bus


def joint_state_to_ros(bus):
    from sensor_msgs.msg import JointState as RosMsg

    out = RosMsg()
    out.header = header_to_ros(bus.header)
    out.name = [str(x) for x in bus.name]
    out.position = list(bus.position)
    out.velocity = list(bus.velocity)
    out.effort = list(bus.effort)
    return out


class SensorMsgsJointStateMapper:
    _ros_type = None

    def ros_msg_type(self):
        cls = type(self)
        if cls._ros_type is None:
            from sensor_msgs.msg import JointState as RosMsg

            cls._ros_type = RosMsg
        return cls._ros_type

    def ros_to_bus(self, msg) -> bytes:
        return joint_state_to_bus(msg).SerializeToString()

    def bus_to_ros(self, payload: bytes):
        BusMsg = _bus_cls()
        bus = BusMsg()
        bus.ParseFromString(payload)
        return joint_state_to_ros(bus)
